storage-node/src/cache/data_store_cache/memdisk/cache_manager.rs

```
@@ -0,0 +1,361 @@
+        if let Some(evicted_bytes_to_disk) = evicted_bytes_to_disk {
+            for (remote_location_evicted, (bytes_vec, data_size)) in evicted_bytes_to_disk {
+                assert_ne!(remote_location_evicted, remote_location);
+                let disk_store_key = self.disk_store.data_store_key(&remote_location_evicted);
+                self.disk_store
+                    .write_data(disk_store_key.clone(), Some(bytes_vec), None)
+                    .await?;
+                let mut disk_replacer = self.disk_replacer.write().await;
+                if disk_replacer
+                    .put(remote_location_evicted, (disk_store_key.clone(), data_size))
+                    .is_none()
+                {
+                    self.disk_store.clean_data(&disk_store_key).await?;
+                }
+            }
+        }
+
+        // 3. If the data is successfully written to memory, directly return.
+        if self.mem_store.is_some() && bytes_to_disk.is_none() {
+            return Ok(bytes_mem_written);
+        }
+
+        // 4. If the data is not written to memory_cache successfully, then cache it to disk.
+        // Need to write the data to network for the current key.
+        let disk_store_key = self.disk_store.data_store_key(&remote_location);
+        // TODO: Write the data store cache first w/ `incompleted` state and update the state
+        // after finishing writing into the data store.
+        let data_size = self
+            .disk_store
+            .write_data(disk_store_key.clone(), bytes_to_disk, Some(data_stream))
+            .await?;
+        let mut disk_replacer = self.disk_replacer.write().await;
+        if disk_replacer
+            .put(remote_location, (disk_store_key.clone(), data_size))
+            .is_none()
+        {
+            self.disk_store.clean_data(&disk_store_key).await?;
+            // TODO: do we need to notify the caller this failure?
+        }
```

Yes. I think we should return an error here.

---

Sure, I will improve it in the future.